Feature/add/some/kind/of/indicator

kira1228 opened the pull request:
Before submitting a pull request (PR), please read the [contributing guide](https://github.com/brainglobe/.github/blob/main/CONTRIBUTING.md).

Please fill out as much of this template as you can, but if you have any problems or questions, just leave a comment and we will help out :)

## Description
Added some kind of indication to show that data is being loaded or uploaded in the background

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
For nicer UX, the indication to show that data is being loaded is useful 

**What does this PR do?**
Introduced an indicator to show how much data is being downloaded.

<img width="1321" alt="Screenshot 2025-03-04 at 2 15 17" src="https://github.com/user-attachments/assets/7672f8a8-8a24-477e-8979-fe5f8f5d31c2" />

## References

Please reference any existing issues/PRs that relate to this PR.

- [issue 112](https://github.com/brainglobe/brainrender-napari/issues/112)

## How has this PR been tested?

Please explain how any new code has been tested, and how you have ensured that no existing functionality has changed.

I tested it by running the following command using the new test code

`pytest -v --color=yes --cov=brainrender_napari --cov-report=xml`

## Is this a breaking change?

If this PR breaks any existing functionality, please explain how and why.
No

## Does this PR require an update to the documentation?
Yes

If any features have changed, or have been added. Please explain how the documentation has been updated (and link to the associated PR). See [here](https://brainglobe.info/community/developers/index.html#to-improve-the-documentation) for details.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality (unit & integration)
- [ ] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

alessandrofelder commented on brainrender_napari/widgets/atlas_manager_view.py:
```diff
@@ -71,15 +105,47 @@ def _on_row_double_clicked(self):
-        worker.returned.connect(self.update_atlas_confirmed.emit)
+        pbar = progress(total=0, desc=f"Updating {atlas_name}...")
+
+        def update_fn(completed, total):
```
I think this function should show the nearest GB/MB (depending on size of the atlas) in the progress bar - currently it displays bytes, like 1456789/4568889, which is not useful for users.

kira1228 replied:
yhea, I think so too. I will fix this in the next PR of brainrender-napari.

alessandrofelder commented on brainrender_napari/widgets/atlas_manager_view.py:
```diff
@@ -90,12 +156,12 @@ def selected_atlas_name(self) -> str:
         selected_atlas_name = self.model().data(selected_atlas_name_index)
         return selected_atlas_name
 
+    @staticmethod
```
could you justify why you made this a `@staticmethod`?

kira1228 replied:
It is defined as a staticmethod because it does not depend on the state of the instance, but only on the function and arguments passed. I thought it would be good as this allows it to be called directly from the class name, improving reusability and ease of testing.

alessandrofelder replied:
(Either way, this change should probably go in a separate PR, as it's unrelated to the progress indicator feature)

kira1228 replied:
Understood, I modified it to an instance method!

alessandrofelder reviewed (changes requested):
Thanks for this @kira1228 - I like how you open small PRs, and show screenshots of them working: this is very helpful.

This effort has allowed us to understand more deeply what is needed to implement this well.

I think this needs to be implemented at a lower level: You may have noticed that you have essentially duplicated the contents of `install_atlas` and `update_atlas` in the code here. To reduce future maintainance work, could you [make a PR to `brainglobe-atlasapi` instead](https://github.com/brainglobe/brainglobe-atlasapi), that allows us to pass the `fn_update` function as an optional argument to `install_atlas/update_atlas` functions?

I can then make a new release of the atlas API.

Then, in `brainrender-napari` you can make a PR with
- [ ] changing `pyproject.toml` to depend on newest `brainglobe-atlasapi
- [ ] implement the changes you propose here with a lot fewer lines of code

kira1228 commented:
Thanks for reviewing my PR! And I've created a [PR](), which addresses above-mentioned problems.

alessandrofelder reviewed (commented):
@kira1228 thanks to your contribution in https://github.com/brainglobe/brainglobe-atlasapi/pull/519 we can now proceed with the improvements :) 

We should simplify the code and change `pyproject.toml` to depend on `brainglobe-atlasapi>=2.2.0`

kira1228 commented:
@alessandrofelder I have implemented following things!

- [x]  changing pyproject.toml to depend on newest brainglobe-atlasapi (>=2.2.0).
- [x]  implement the changes I propose here with a lot fewer lines of code.
- [x]  Change _apply_in_thread function to an instance method.

If I need futher improvements, please let me know 👍

kira1228 commented:
@alessandrofelder 

Thank for replying and it's totally fine to use Qt widget instead of napari plugin.

All right, for better UX, I will implement indicator using `qtpy.QtWidgets.QProgressbar`

kira1228 commented:
@alessandrofelder 

Hi, I've implemented download/update progress bar using `qtpy.QtWidgets.QProgressbar `! 

If I need futher improvements, please let me know 😄

alessandrofelder commented on brainrender_napari/widgets/atlas_manager_view.py:
```diff
@@ -19,16 +19,48 @@
 from brainrender_napari.widgets.atlas_manager_dialog import AtlasManagerDialog
 
 
+def _format_bytes(num_bytes: float) -> str:
+    for unit in ["B", "KB", "MB", "GB"]:
+        if num_bytes < 1024:
+            return f"{num_bytes:.2f} {unit}"
+        num_bytes /= 1024
+    return f"{num_bytes:.2f} TB"
```
This would fit well into `utils/formatting.py` as a public function (+tests for it)

kira1228 replied:
I think so too, I will implement in that way 👍

alessandrofelder commented on brainrender_napari/widgets/atlas_manager_view.py:
```diff
@@ -19,16 +19,48 @@
+    return f"{num_bytes:.2f} TB"
+
+
+def install_atlas_with_progress(atlas_name: str, fn_update: Callable):
+    """
+    By passing fn_update when instantiating BrainGlobeAtlas
+    Progress information during the downlaod progress is
+    retrieved so that the actual progress (%) can be updated
+    """
+    install_atlas(atlas_name, fn_update=fn_update)
+    return atlas_name
+
+
+def update_atlas_with_progress(atlas_name: str, fn_update: Callable):
+    """
+    When updating an existing atlas, it should also be possible to pass
+    fn_update to get the actual progress (%)
+    """
+    update_atlas(atlas_name, fn_update=fn_update)
+    return atlas_name
```
It feels like this functions don't add much, could we just call the `brainglobe_atlasapi` functions directly where we need these? But maybe I am missing something?

alessandrofelder commented on brainrender_napari/widgets/atlas_manager_view.py:
```diff
@@ -19,16 +19,48 @@
 class AtlasManagerView(QTableView):
     download_atlas_confirmed = Signal(str)
     update_atlas_confirmed = Signal(str)
+    progress_updated = Signal(int, int, str, object)
```
Yes, this is the way to go, I think - very nice!

alessandrofelder reviewed (changes requested):
Thanks a lot for this big effort @kira1228 - The functionality is great and exactly what we want. But I think the software design could be much improved:

A software design suggestion, to improve separation of concerns and modularity:
* the `AtlasManagerView` should not have a progress bar, it just emit the `progress_updated` signal, and not care whether anyone is listening for this
* the progress bar should be its own (minimal) class
* the `BrainrenderManagerWidget` should have instances of both these classes, and connect the `progress_updated` signal from one to a function in the other.

Also, we expect some tests for the added functionality.

Thanks for your patience with my (very high) expectations - let me know if you need more in-depth explanations/help with any of my suggestions.

kira1228 commented:
Hi @alessandrofelder 

Based on your advice, I've implemented the following things:

- [x] Remove the progress bar handling from `AtlasManagerView`

- [x] Create a progress bar widget class 

- [x] Connect signals in the BrainrenderManagerWidget class

- [x] Move the` _format_bytes function` to `utils/formatting.py`

- [x] Eliminate the `install_atlas_with_progress/update_atlas_with_progress` functions that don't add much value 
- [x] changed parts of the test code to accommodate this change. (Not all test codes have been implemented yet.)

kira1228 commented:
I am considering running the following tests in addition to the above changes.

- In `test_atlas_manager_view.py`, add code to check if the progress bar signals are being sent out

- In `test_atlas_progress_bar.py` test the AtlasProgressBar class to check the progress bar behavior.

- In `test_brainrender_manager_widget.py`, test the signal connection of BrainrenderManagerWidget.

Can I implement the test code in this way?

alessandrofelder commented:
Hey @kira1228 thanks for this

I may not have time to review it deeply this week, but here's a start:
* I've done some manual testing: can you make the progress bar display when an atlas is updated as well?
* your suggested test strategy sounds good in general

> Also, the test is failing on ubuntu-latest py3.11/3.12, resulting in FileNotFoundError: [Errno 2] No such file or directory: '/home/runner/.brainglobe/example_mouse_100um_v1.2'
>Can I solve this problem by completely removing the actual file system operations (shutil.rmtree, atlas_directory.iterdir(), Path.exists(file), etc.) in test_download_confirmed_callbackso so that I can mock the required behavior independent of the actual file system operations and limit the scope of the test to confirm that the signals are issued correctly?

yes, this is sensible in principle.

kira1228 commented:
Hi @alessandrofelder

Thanks for the reply!

And I haven't pushed all my intended changes yet. But I will do the following things

- make the progress bar display when an atlas is updated as well

- Add some test code as I mentioned above

kira1228 commented:
Hi, @alessandrofelder 

I've implemented some test cases and verified that the progress bar is displayed when updating through the `test_different_operation_types` function. I also mocked some code to ensure that the code doesn't depend on a specific filesystem.

If I need futher improvements, please let me know! 👍

alessandrofelder reviewed (changes requested):
Thanks @kira1228 - in general, it's very good but there's still some problems
* have you tried the functionality manually? It works well for me when I download and atlas but not when I update it

I also have some more questions about your tests, and I think one of the tests doesn't actually test the source code: I made a suggestion on how to improve that one.

Thanks for your patience!

kira1228 commented:
@alessandrofelder - Thanks for your rapid reply!

- have you tried the functionality manually? It works well for me when I download and atlas but not when I update it

I have confirmed through a test function that the update signal is being emitted, but since there is no outdated atlas, I haven't verified it visually. Is there a way to prepare an outdated atlas for testing purposes?

And thanks for your suggestions!

kira1228 commented on brainrender_napari/widgets/atlas_progress_bar.py:
```diff
@@ -0,0 +1,55 @@
+"""
+A progress bar widget specifically for atlas download and update operations.
```
When I ran the test, I encountered an error that was dependent on the file path. Since I didn't want to modify `mock_newer_atlas_version_available` to avoid depending on the file system, I used `mocker` instead to make the test independent of the file system.

kira1228 replied:
Initially, I placed it in `brainrender-napari/widgets/atlas_progress_bar`, but later moved it to `brainrender-napari/atlas_progress_bar`, which seems to have caused a duplication. I will commit later to place it in `widgets/atlas_progress_bar`.

kira1228 commented on tests/test_unit/test_atlas_manager_view.py:
```diff
@@ -141,62 +179,102 @@ def test_hover_atlas_manager_view(atlas_manager_view, mocker):
 
 
-def test_get_tooltip_invalid_name():
+def test_get_tooltip_invalid_name(mocker):
```
When I ran the test, I encountered an error that was dependent on the file path. Since I didn't want to modify mock_newer_atlas_version_available to avoid depending on the file system, I used mocker instead to make the test independent of the file system.

kira1228 commented on tests/test_unit/test_atlas_manager_view.py:
```diff
@@ -35,6 +34,32 @@ def test_update_atlas_confirmed(
     local_version_index = atlas_manager_view.model().index(0, 2)
     assert atlas_manager_view.model().data(local_version_index) == "1.1"
 
+    def fake_update_atlas(atlas_name, fn_update):
```
Sorry, I may have misunderstood the intention of your comment.

Is my understanding correct that you are suggesting using `mock_newer_atlas_version_available` instead of the `fake_update_atlas` function to recreate the same situation?

kira1228 replied:
I've implemented `test_update_atlas_confirmed` code using `mock_newer_atlas_version_available` like the code snippet below.

```python
def test_update_atlas_confirmed(
    qtbot,
    mocker,
    atlas_manager_view,
    mock_newer_atlas_version_available,
):
    """
    Test that confirming atlas update triggers the correct actions
    and emits signal.
    Uses filesystem fixture to simulate an outdated atlas.
    """
    mocker.patch.object(
        atlas_manager_view,
        "selected_atlas_name",
        return_value="example_mouse_100um",
    )
    
    mocker.patch(
        "brainrender_napari.widgets.atlas_manager_view.update_atlas",
        return_value="example_mouse_100um",
    )

    with qtbot.waitSignal(
        atlas_manager_view.update_atlas_confirmed,
        timeout=300000,  # assumes atlas can be updated in 5 minutes!
    ) as update_atlas_confirmed_signal:
        atlas_manager_view._on_update_atlas_confirmed()

    assert update_atlas_confirmed_signal.args == ["example_mouse_100um"]
``` 

However, when I ran this test function, I got an `AssertionError` saying that `Path.home() / ".brainglobe/example_mouse_100um_v1.2"` does not exist, while the actual path exists at `/Users/kira/.brainglobe-tests/.brainglobe/example_mouse_100um_v1.2.` To resolve this issue, I thought it would be better to use mocks to make the design independent of the file path. Do you have any other suggestions for a better approach?

kira1228 commented:
Oh, Ok 😄 

Thanks for the trick, I will do that!

kira1228 commented:
Hi, @alessandrofelder 

I have modified the `_on_update_atlas_confirmed` function so that the progress bar's progress is properly reflected during the update. I also added a new image in the PR. 

However, since I am not using the `update_atlas` function and the code became a bit more complex, please let me know if this is an unintended change. 

When I tested manually using the `update_atlas` function, it worked as expected in the terminal, but the `QtWidget` did not reflect it consistently. I couldn't figure out why this was happening, so the current implementation is what I came up with.

kira1228 commented:
Also, when the atlas is updated, the local version is not changed within the `Atlas Manager view` because the `AtlasTableModel` only fetches data once during initialization and is not automatically updated afterward. 

I thought it would be better to update the atlas table when the version is updated. What do you think about this?


```python
    def _on_update_atlas_confirmed(self):
        """Updates the currently selected atlas and signals this."""
        atlas_name = self.selected_atlas_name()
        self._start_worker(
            self._wrapped_update_atlas,
            atlas_name,
            self.update_atlas_confirmed,
            "Updating",
        )
    
        # Connect update completion signals to table update process
        if self.receivers(self.update_atlas_confirmed) == 1:  
            self.update_atlas_confirmed.connect(self._refresh_table_after_update)

    def _refresh_table_after_update(self, atlas_name):
        """Update table view after atlas update"""
        # Reload the model to reflect the latest information
        self.setModel(AtlasTableModel(AtlasManagerView))
        self.resizeColumnsToContents()
        self.hideColumn(self.model().column_headers.index("Raw name"))
    
        # select updated atlas row
        self._select_atlas_by_name(atlas_name)

    def _select_atlas_by_name(self, atlas_name):
        """
        Select the atlas with the given name.
        """
        for row in range(self.model().rowCount()):
            index = self.model().index(row, 0) 
            if self.model().data(index) == atlas_name:
                self.setCurrentIndex(index)
                break

    def _wrapped_update_atlas(self, atlas_name, fn_update=None):
        """Wrapper around update_atlas to ensure progress updates are properly handled."""
        if fn_update:
            fn_update(0, 100)
    
        atlas = BrainGlobeAtlas(atlas_name=atlas_name, check_latest=False)
    
        if atlas.check_latest_version(print_warning=False):
            if fn_update:
                fn_update(100, 100)  # notify completed
            return atlas_name
    
        # remove old atlas
        fld = atlas.brainglobe_dir / atlas.local_full_name
        shutil.rmtree(fld)
    
        BrainGlobeAtlas(atlas_name=atlas_name, fn_update=fn_update)
    
        if fn_update:
            fn_update(100, 100)
        
        return atlas_name
```

alessandrofelder reviewed (commented):
# Tests + tidying up

Thanks @kira1228 - looking good!

It looks like the tests don't pass (it's not an internet issue this time, I think, and looks connected to your mocks), and there is still the duplicate `atlas_progressbar.py` file? We need to tidy that and get the tests to pass before merging - almost there!!

This is the main thing that is missing, then I will merge!

# Extras (for later, not in this PR)

### the extra function

Your solution with
```python
def _wrapped_update_atlas(self, atlas_name, fn_update=None):
```
works well, but it is a bit "hacky". It would be better if we found out why `update_atlas` doesn't work the same way as `install_atlas`??? Maybe it points to a problem in the `brainglobe_atlasapi` that we should know about? But let's solve that in a different issue #189


### refreshing the data in the view
> I thought it would be better to update the atlas table when the version is updated. What do you think about this?

[This PR](https://github.com/brainglobe/brainrender-napari/pull/184/files) has already a nice solution to this, so we do not need to worry about that here.

For explanation, I _think_ (:crossed_fingers: :joy: ) you can do this more simply:

Instead of this function
```python
def _refresh_table_after_update(self, atlas_name):
     ...
```

I can just do
```
self.model().refresh_data()
```

This will save you a function and will also be more performant. It should be reflected in the view automatically - this is the whole point of the model-view pattern :)

kira1228 commented:
@alessandrofelder Thanks for the reply!

I have figured out the cause of the issue where the atlas progress bar is not displayed during the update.

The current `_start_worker` method calls both `install_atlas` and `update_atlas` in the same way, but these functions have different argument orders:


```Python
install_atlas(atlas_name, fn_update=None)
update_atlas(atlas_name, force=False, fn_update=None)
```

As a result, the callback function for tracking progress is correctly passed as `fn_update` during the download, but during the update, it is mistakenly passed as the force parameter, leaving `fn_update` as None.

To fix this, I made sure that the appropriate arguments are passed based on the operation type, and I was able to confirm that the progress bar now displays correctly during both the update and download.

```Python
    def _start_worker(
        self,
        operation: Callable,
        atlas_name: str,
        signal: Signal,
        operation_type: str,
    ):
        def update_fn(completed, total):
            self.progress_updated.emit(
                completed, total, atlas_name, operation_type
            )
        
        if operation.__name__ == "update_atlas":
            worker = self._apply_in_thread(operation, atlas_name, fn_update=update_fn)
        else:
            worker = self._apply_in_thread(operation, atlas_name, update_fn)

        worker.returned.connect(lambda result: signal.emit(result))
        worker.start()

```
This change was made without modifying `brainglobe_atlasapi` and was done purely to verify that this hypothesis is correct. As suggested in the [PR](https://github.com/brainglobe/brainglobe-atlasapi/pull/539) I think it would be better to change the argument order in `brainglobe_atlasapi`.

Therefore, could you please check the [PR](https://github.com/brainglobe/brainglobe-atlasapi/pull/539) and release a new version with the updated argument order?

Then, we can focus on creating the test code and addressing the duplication issue with `atlas_progress_bar`.

alessandrofelder commented:
Very good detective work, @kira1228 - thanks a lot for digging into it.

I think we can get away without changing the `brainglobe-atlasapi` actually - we can pass `fn_update` argument as a keyword instead of as a positional argument to `_apply_in_thread`.

Sharing [an intro to positional and keyword arguments in Python](https://thepythoncodingbook.com/2022/11/11/positional-arguments-and-keyword-arguments-in-python/) in case it's helpful.

kira1228 commented:
@alessandrofelder 

Ah, you're right, I somehow didn't notice that. 🤣 
I implemented it using that method. Also, I've changed the path for `atlas_progress_bar.py.`

Please let me know if any further modifications are necessary!

alessandrofelder commented on brainrender_napari/widgets/atlas_manager_view.py:
```diff
@@ -68,19 +74,49 @@ def _on_row_double_clicked(self):
             )
             download_dialog.exec()
 
+    def _start_worker(
```
very nitpicky: I think this function obfuscates what's happening slightly, for not much reduction in code duplication. I suggest to tackle this in a separate issue #190

kira1228 replied:
I will do that once I submit my proposal 👍

alessandrofelder commented on brainrender_napari/widgets/atlas_manager_view.py:
```diff
@@ -68,19 +74,49 @@ def _on_row_double_clicked(self):
+        operation_type: str,
+    ):
+        """
+        Helper function that combines progress bar generation,
+        update processing, worker activation, and signal issuance.
+        Displays a QProgressBar in the plugin widget.
```
```suggestion
        Helper function that connects the `progress_updated` signal to the underlying atlas API progress update function, and then starts a download/update operation in separate thread, ensuring it returns `signal` when the thread operation finishes.
        
```

alessandrofelder reviewed (approved):
This is great @kira1228 - very useful feature and the code quality is now sufficient for merging :tada: 

I made a follow-on issue with a suggested refactor, for a minor detail #190
